src/core/lib/event_engine/cf_engine/cfstream_endpoint.cc

```diff
@@ -0,0 +1,343 @@
+          DoRead(std::move(on_read), buffer);
+        } else {
+          on_read(status);
+        }
+      },
+      false /* is_permanent*/));
+
+  return false;
+}
+
+void CFStreamEndpoint::DoRead(absl::AnyInvocable<void(absl::Status)> on_read,
+                              SliceBuffer* buffer) {
+  GRPC_EVENT_ENGINE_ENDPOINT_TRACE("CFStreamEndpoint::DoRead, this: %p", this);
+
+  int buffer_size = 8192;
+  auto buffer_index =
+      buffer->AppendIndexed(Slice(memory_allocator_.MakeSlice(buffer_size)));
+
+  CFIndex read_size = CFReadStreamRead(
+      cf_read_stream_,
+      internal::SliceCast<MutableSlice>(buffer->MutableSliceAt(buffer_index))
+          .begin(),
+      buffer_size);
+
+  if (read_size < 0) {
+    auto status = CFErrorToStatus(CFReadStreamCopyError(cf_read_stream_));
+    GRPC_EVENT_ENGINE_ENDPOINT_TRACE("CFStream read error: %s",
+                                     status.ToString().c_str());
+    on_read(status);
+    return;
+  }
+
+  buffer->RemoveLastNBytes(buffer->Length() - read_size);
+
+  if (grpc_event_engine_endpoint_data_trace.enabled()) {
+    for (size_t i = 0; i < buffer->Count(); i++) {
+      auto str = buffer->RefSlice(i).as_string_view();
+      gpr_log(GPR_INFO, "CFStreamEndpoint::DoRead, this: %p (peer=%s): %.*s",
+              this, peer_address_string_.c_str(),
+              static_cast<int>(str.length()), str.data());
+    }
+  }
+
+  on_read(absl::OkStatus());
+}
+
+bool CFStreamEndpoint::Write(absl::AnyInvocable<void(absl::Status)> on_writable,
+                             SliceBuffer* data, const WriteArgs* /* args */) {
+  write_event_.NotifyOn(new PosixEngineClosure(
+      [this, on_writable = std::move(on_writable),
+       data](absl::Status status) mutable {
+        if (status.ok()) {
+          DoWrite(std::move(on_writable), data);
+        } else {
+          on_writable(status);
+        }
+      },
+      false /* is_permanent*/));
+
+  return false;
+}
+
+void CFStreamEndpoint::DoWrite(
+    absl::AnyInvocable<void(absl::Status)> on_writable, SliceBuffer* data) {
+  GRPC_EVENT_ENGINE_ENDPOINT_TRACE("CFStreamEndpoint::DoWrite, this: %p", this);
+  if (grpc_event_engine_endpoint_data_trace.enabled()) {
+    for (size_t i = 0; i < data->Count(); i++) {
+      auto str = data->RefSlice(i).as_string_view();
+      gpr_log(GPR_INFO, "CFStreamEndpoint::DoWrite, this: %p (peer=%s): %.*s",
+              this, peer_address_string_.c_str(),
+              static_cast<int>(str.length()), str.data());
+    }
+  }
+
+  size_t total_written_size = 0;
+  for (size_t i = 0; i < data->Count(); i++) {
+    auto slice = data->RefSlice(i);
+    size_t written_size =
+        CFWriteStreamWrite(cf_write_stream_, slice.begin(), slice.size());
+
+    total_written_size += written_size;
+    if (written_size < slice.size()) {
+      SliceBuffer written;
+      data->MoveFirstNBytesIntoSliceBuffer(total_written_size, written);
+
+      write_event_.NotifyOn(new PosixEngineClosure(
```

Comment:
Am I reading it correctly, that this is recursively scheduling a write, if it cannot finish in one `CFWriteStreamWrite` call?

I wonder if this is susceptible to a race condition if `CFStreamEndpoint::Write` is called again and now there are 2 scheduled writes?

Reply:
"There can be at most one outstanding write per Endpoint at any given time", this is how c-core uses endpoint.

Reply:
aha ok got it! Good to note! Thanks for clarifying!